web/impact/impact/v1/views/cancel_office_hour_session_view.py

```diff
@@ -0,0 +1,126 @@
+PERMISSION_DENIED = 'Office hour session could not be canceled. ' \
+                    'You do not have permission to cancel that session'
+
+
+def get_office_hours_list_url(family_slug, program_slug):
+    site_url = 'accelerate.masschallenge.org'
+    return f'https://{site_url}/officehours/list/{family_slug}/{program_slug}'
+
+
+def get_office_hour_shared_context(office_hour, message=None):
+    family_slug = office_hour.program.program_family.url_slug
+    program_slug = office_hour.program.url_slug
+    tz = timezone(office_hour.location.timezone or DEFAULT_TIMEZONE)
+    date = office_hour.start_date_time.astimezone(tz).strftime('%A, %d %B, %Y')
+    start_time = office_hour.start_date_time.astimezone(tz).strftime('%I:%M%p')
+    end_time = office_hour.end_date_time.astimezone(tz).strftime('%I:%M%p')
+    return {
+        'date': date,
+        'start_time': start_time,
+        'end_time': end_time,
+        'location': office_hour.location.name if office_hour.location else '',
+        'dashboard_url': get_office_hours_list_url(family_slug, program_slug),
+        'mentor_name': office_hour.mentor.get_profile().full_name(),
+        'phone': office_hour.program.program_family.phone_number,
+        'message': message,
+    }
+
+
+def get_ui_notification(context, staff=False):
+    if staff:
+        return STAFF_NOTIFICATION.format(**context)
+    return MENTOR_NOTIFICATION.format(**context)
+
+
+class CancelOfficeHourSessionView(ImpactView):
+    permission_classes = (OfficeHourPermission,)
+    view_name = 'cancel_office_hour_session_view'
+
+    def cancel_office_hour_session(self, office_hour, user, message):
+        shared_context = get_office_hour_shared_context(office_hour, message)
+        if user == office_hour.mentor:
+            if office_hour.finalist:
+                raise PermissionDenied(PERMISSION_DENIED)
+            self.handle_notification(office_hour, shared_context, 'You have')
+            ui_notification = get_ui_notification(shared_context)
+            office_hour.delete()
+            return ui_notification
+        else:
+            self.handle_notification(
+                office_hour, shared_context, 'MassChallenge has')
+            ui_notification = get_ui_notification(shared_context, staff=True)
+            office_hour.delete()
+            return ui_notification
+
+    def send_email(self, context, to_addr):
+        html_email = loader.render_to_string(
+            'emails/cancel_office_hour_session_email.html',
+            context
+        )
+        MinimalEmailHandler(
+            to=to_addr,
+            subject=SUBJECT.format(**context),
+            body=None,
+            from_email=settings.NO_REPLY_EMAIL,
+            attach_alternative=[html_email, 'text/html'],
+        ).send()
+
+    def post(self, request):
+        message = request.data.get('message', None)
+        id = request.data.get('id', None)
+        office_hour = get_object_or_404(MentorProgramOfficeHour, id=id)
```

Comment:
I think this will throw an error if a user commits a double-tap when they post their request. (since there will be two requests, and the second one will find the office hour has been deleted and throw a 404)

This is probably not ideal UX. 

I would say, if no office hour with that ID exists, we should either consider that a success (they wanted it deleted, it's deleted...) or report back something informative like "You wanted to delete that office hour, but unfortunately, it doesn't exist, so we can't... um... delete it. Right. Nevermind, we're good, bye!"

Reply:
I agree we should let the user delete an office hour with a nonexistent ID if it is theirs to delete. There is no need to display a UI message in this case. However, if we are concerned there may be instances of non-existing office hour ID's, would it make sense to have some logging for our records at the very least?

Reply:
@hfreund the issue is, if a user submits a request with an id that does not correspond to any existing office hour, then the most likely scenario is that they've tried to delete a session that has already been deleted. In this case, we'd rather not error out on them, because probably they'll have just double-tapped the submit button. 

Anyway, this ship has sailed. If you have questions about the existing behavior, ping me and we can get you answers.

Reply:
Nope. I think we're in agreement here.